File: packages/typescript/src/span/builder.ts
/**
 * SpanBuilder — open/close + hash-chain finalization.
 *
 * Mirrors agent_capture.span.builder. Children buffer with their parent
 * until the parent closes; parent then stamps every pending descendant
 * with its content_hash and ships the subtree leaves-first. Root ships
 * last. This is the only ordering consistent with "compute provenance as
 * the last step before export" while keeping content_hash a function of
 * the span's own content alone.
 */

import { randomBytes } from "node:crypto";

import { ErrorCode, logError } from "../internal/safelog.js";
import { currentParent } from "../context/propagation.js";
import type { SpanExporter } from "../exporter/base.js";
import { contentHash } from "../schema/canonical.js";
import { SCHEMA_VERSION } from "../schema/version.js";
import type {
  AgentCaptureSpan,
  ComplianceMetadata,
  ErrorInfo,
  Attributes as TypedAttributes,
} from "../schema/span.js";

import type { OpenSpan, SpanStatus, SpanType } from "./types.js";

function newTrajectoryId(): string {
  return randomBytes(16).toString("hex");
}

function newSpanId(): string {
  return randomBytes(8).toString("hex");
}

/** Format a Date as "YYYY-MM-DDTHH:MM:SS.ffffffZ" matching canonical.ts. */
function isoCanonical(d: Date): string {
  const pad = (n: number, w: number) => String(n).padStart(w, "0");
  return (
    `${pad(d.getUTCFullYear(), 4)}-${pad(d.getUTCMonth() + 1, 2)}-${pad(d.getUTCDate(), 2)}T` +
    `${pad(d.getUTCHours(), 2)}:${pad(d.getUTCMinutes(), 2)}:${pad(d.getUTCSeconds(), 2)}.` +
    `${pad(d.getUTCMilliseconds(), 3)}000Z`
  );
}

export interface OpenOptions {
  name: string;
  type: SpanType;
  attributes: TypedAttributes;
  compliance?: ComplianceMetadata;
  inputs?: unknown;
  /** Explicit parent override (framework adapters that wire by run_id). */
  parent?: OpenSpan | null;
}

export interface CloseOptions {
  outputs?: unknown;
  status?: SpanStatus;
  error?: ErrorInfo;
}

export interface BuilderOptions {
  defaultCompliance?: ComplianceMetadata;
  redactionFilter?: { apply(span: AgentCaptureSpan): AgentCaptureSpan };
}

export class SpanBuilder {
  constructor(
    private readonly exporter: SpanExporter,
    private readonly options: BuilderOptions = {},
  ) {}

  open({
    name,
    type,
    attributes,
    compliance,
    inputs,
    parent,
  }: OpenOptions): OpenSpan {
    const effectiveCompliance = compliance ?? this.options.defaultCompliance;
    if (effectiveCompliance === undefined) {
      throw new Error(
        "No compliance metadata available. Pass compliance or configure({ defaultCompliance })",
      );
    }
    const resolvedParent = parent ?? currentParent();
    const trajectoryId =
      resolvedParent === null ? newTrajectoryId() : resolvedParent.trajectoryId;
    const spanId =
      resolvedParent === null ? trajectoryId.slice(0, 16) : newSpanId();
    const parentSpanId = resolvedParent === null ? null : resolvedParent.spanId;
    return {
      spanId,
      parentSpanId,
      trajectoryId,
      name,
      type,
      startTime: new Date(),
      attributes,
      compliance: effectiveCompliance,
      inputs: inputs ?? null,
      parent: resolvedParent,
      pending: [],
    };
  }

  close(openSpan: OpenSpan, { outputs, status, error }: CloseOptions = {}): AgentCaptureSpan | null {
    try {
      return this.closeImpl(openSpan, outputs ?? null, status ?? "ok", error ?? null);
    } catch (exc) {
      logError(ErrorCode.AC101, "SpanBuilder.close failed", exc);
      return null;
    }
  }

  private closeImpl(
    openSpan: OpenSpan,
    outputs: unknown,
    status: SpanStatus,
    error: ErrorInfo | null,
  ): AgentCaptureSpan {
    let finalized: AgentCaptureSpan = {
      span_id: openSpan.spanId,
      parent_span_id: openSpan.parentSpanId,
      trajectory_id: openSpan.trajectoryId,
      name: openSpan.name,
      type: openSpan.type,
      start_time: isoCanonical(openSpan.startTime),
      end_time: isoCanonical(new Date()),
      status,
      error,
      inputs: openSpan.inputs,
      outputs,
      attributes: openSpan.attributes,
      compliance: openSpan.compliance,
      provenance: {
        content_hash: "0".repeat(64),
        parent_content_hash: null,
        schema_version: SCHEMA_VERSION,
      },
    } as AgentCaptureSpan;

    // Redaction runs BEFORE hashing so content_hash covers shipped bytes.
    if (this.options.redactionFilter !== undefined) {
      finalized = this.options.redactionFilter.apply(finalized);
    }

    // content_hash excludes provenance, so the placeholder above is fine.
    const myHash = contentHash(finalized);
    finalized = {
      ...finalized,
      provenance: {
        content_hash: myHash,
        parent_content_hash: null,
        schema_version: SCHEMA_VERSION,
      },
    };

    // Stamp every pending descendant with my hash, then ship them.
    for (const child of openSpan.pending) {
      const stamped: AgentCaptureSpan = {
        ...child,
        provenance: {
          content_hash: child.provenance.content_hash,
          parent_content_hash: myHash,
          schema_version: child.provenance.schema_version ?? SCHEMA_VERSION,
        },
      };
      this.exporter.export(stamped);
    }
    openSpan.pending = [];

    if (openSpan.parent === null) {
      // Root: ship directly.
      this.exporter.export(finalized);
    } else {
      // Non-root: park with parent. Parent will stamp parent_content_hash on close.
      openSpan.parent.pending.push(finalized);
    }
    return finalized;
  }
}

File: packages/python/src/agent_capture/span/builder.py
"""Span builder — constructs and finalizes spans across their lifecycle.

Lifecycle:

1. ``builder.open(...)`` returns a mutable :class:`OpenSpan`. The caller
   wraps its work in :func:`~agent_capture.context.propagation.span_scope`
   so the OpenSpan becomes the current parent for any inner spans.
2. Work happens. Inner spans open and close, registering themselves as
   pending children of their parent.
3. ``builder.close(open_span, outputs=...)`` finalizes the span: validates
   it against the Pydantic schema, computes its ``content_hash`` over the
   canonical serialization, and decides what to ship.

Hash-chain ordering (architecture §4.4): each span's ``content_hash`` is
SHA-256 of its own canonical content (provenance excluded). ``parent_content_hash``
is the parent's final ``content_hash``. Because parent's hash isn't known
until parent closes (parent closes *after* its children), closing children
**buffer** with their parent. When the parent closes, it stamps every
pending descendant with its own ``content_hash`` and ships the whole
subtree to the exporter in leaf-first order. The root span ships last.

This is the only correct interpretation under the architecture's "compute
provenance as the last step before export" rule.
"""

from __future__ import annotations

import secrets
from dataclasses import dataclass, field
from datetime import UTC, datetime
from typing import Any

from agent_capture._internal.safelog import ErrorCode, log_error
from agent_capture.context.propagation import current_parent
from agent_capture.errors import SchemaViolation
from agent_capture.exporter.base import SpanExporter
from agent_capture.redaction.filter import RedactionFilter
from agent_capture.schema import (
    ComplianceMetadata,
    ErrorInfo,
    ProvenanceFields,
    Span,
    SpanStatus,
    SpanType,
)
from agent_capture.schema.canonical import content_hash
from agent_capture.schema.types import TypedAttributes


def _new_trajectory_id() -> str:
    """Generate an OTel-compatible 128-bit trace id as 32 hex chars."""
    return secrets.token_hex(16)


def _new_span_id() -> str:
    """Generate an OTel-compatible 64-bit span id as 16 hex chars."""
    return secrets.token_hex(8)


def _now_utc() -> datetime:
    return datetime.now(tz=UTC)


@dataclass
class OpenSpan:
    """Mutable in-progress span state.

    Held by the span builder between ``open(...)`` and ``close(...)``. Lives
    in the contextvar so inner spans can locate their parent.
    """

    span_id: str
    parent_span_id: str | None
    trajectory_id: str
    name: str
    type: SpanType
    start_time: datetime
    attributes: TypedAttributes
    compliance: ComplianceMetadata
    inputs: Any | None = None
    _parent: OpenSpan | None = field(default=None, repr=False)
    _pending: list[Span] = field(default_factory=list, repr=False)


class SpanBuilder:
    """Factory for opening and closing spans.

    A single builder is shared across the process. Thread- and async-safe
    because state lives in the contextvar (for parent tracking) or in the
    exporter (for queueing); the builder itself is stateless aside from
    its references to the exporter and default compliance metadata.
    """

    def __init__(
        self,
        exporter: SpanExporter,
        *,
        default_compliance: ComplianceMetadata | None = None,
        redaction_filter: RedactionFilter | None = None,
        flush_on_root_close: bool = False,
    ) -> None:
        self._exporter = exporter
        self._default_compliance = default_compliance
        self._redaction_filter = redaction_filter
        # Serverless mode: synchronously flush the exporter when a trajectory's
        # root span closes, so spans aren't stranded in the async queue after a
        # Cloud Run / Lambda response returns and CPU is throttled. See #64.
        self._flush_on_root_close = flush_on_root_close

    def open(
        self,
        *,
        name: str,
        type: SpanType,
        attributes: TypedAttributes,
        compliance: ComplianceMetadata | None = None,
        inputs: Any | None = None,
        parent: OpenSpan | None = None,
    ) -> OpenSpan:
        """Open a new span.

        Parent resolution order:

        1. Explicit ``parent=`` argument (used by framework adapters that
           wire parent-child by their own run id rather than via the
           ``contextvars`` parent pointer).
        2. Otherwise, :func:`agent_capture.context.current_parent`.

        Raises:
            SchemaViolation: If no compliance metadata is available (neither
                passed in nor configured as a default). This is *internal*
                — callers wrap with try/except per the never-raises rule.
        """
        effective_compliance = compliance or self._default_compliance
        if effective_compliance is None:
            raise SchemaViolation(
                "No compliance metadata available. Pass compliance=... or configure() a default before opening a span."
            )

        if parent is None:
            parent = current_parent()
        if parent is None:
            trajectory_id = _new_trajectory_id()
            span_id = trajectory_id[:16]
            parent_span_id = None
        else:
            trajectory_id = parent.trajectory_id
            span_id = _new_span_id()
            parent_span_id = parent.span_id

        return OpenSpan(
            span_id=span_id,
            parent_span_id=parent_span_id,
            trajectory_id=trajectory_id,
            name=name,
            type=type,
            start_time=_now_utc(),
            attributes=attributes,
            compliance=effective_compliance,
            inputs=inputs,
            _parent=parent,
        )

    def close(
        self,
        open_span: OpenSpan,
        *,
        outputs: Any | None = None,
        status: SpanStatus = SpanStatus.OK,
        error: ErrorInfo | None = None,
    ) -> Span | None:
        """Finalize the span, compute its hash, and ship the subtree if root.

        Returns the finalized :class:`Span` for callers that want to inspect
        it (tests, etc.). On internal failure, logs to safelog and returns
        ``None`` — never raises into the caller.
        """
        try:
            return self._close_impl(open_span, outputs=outputs, status=status, error=error)
        except Exception as exc:
            log_error(ErrorCode.AC101, "SpanBuilder.close failed: %s", exc)
            return None

    def _close_impl(
        self,
        open_span: OpenSpan,
        *,
        outputs: Any | None,
        status: SpanStatus,
        error: ErrorInfo | None,
    ) -> Span:
        finalized = Span(
            span_id=open_span.span_id,
            parent_span_id=open_span.parent_span_id,
            trajectory_id=open_span.trajectory_id,
            name=open_span.name,
            type=open_span.type,
            start_time=open_span.start_time,
            end_time=_now_utc(),
            status=status,
            error=error,
            attributes=open_span.attributes,
            inputs=open_span.inputs,
            outputs=outputs,
            compliance=open_span.compliance,
            provenance=ProvenanceFields(
                content_hash="0" * 64,  # placeholder; replaced below
                parent_content_hash=None,
            ),
        )

        # Redaction runs BEFORE the hash is computed so content_hash covers
        # the bytes that actually ship. Architecture doc §4.4 + §8: hash
        # the post-redaction canonical form.
        if self._redaction_filter is not None:
            finalized = self._redaction_filter.apply(finalized)

        # content_hash is computed over canonical bytes with provenance excluded,
        # so the placeholder content_hash above doesn't affect the result.
        my_hash = content_hash(finalized)
        finalized = finalized.model_copy(
            update={
                "provenance": ProvenanceFields(
                    content_hash=my_hash,
                    parent_content_hash=None,
                )
            }
        )

        # Stamp every pending descendant under us with our hash, then ship them.
        for child in open_span._pending:
            stamped = child.model_copy(
                update={
                    "provenance": ProvenanceFields(
                        content_hash=child.provenance.content_hash,
                        parent_content_hash=my_hash,
                    )
                }
            )
            self._exporter.export(stamped)
        open_span._pending.clear()

        if open_span._parent is None:
            # Root: nothing to wait on. Ship ourselves directly.
            self._exporter.export(finalized)
            if self._flush_on_root_close:
                self._flush_exporter()
        else:
            # Non-root: park with parent. Parent will stamp parent_content_hash on close.
            open_span._parent._pending.append(finalized)

        return finalized

    def _flush_exporter(self) -> None:
        """Synchronously drain the exporter (serverless mode). Never raises."""
        flush = getattr(self._exporter, "flush", None)
        if flush is None:
            return
        try:
            flush()
        except Exception as exc:
            log_error(ErrorCode.AC416, "flush_on_root_close: exporter.flush raised: %s", exc)
